src/rust/engine/src/nodes.rs

```diff
@@ -372,18 +405,47 @@ pub struct SelectDependencies {
   pub subject: Key,
   pub variants: Variants,
   pub selector: selectors::SelectDependencies,
+  pub dep_product_entries: rule_graph::Entries,
+  pub product_entries: rule_graph::Entries,
 }
 
 impl SelectDependencies {
   pub fn new(
     selector: selectors::SelectDependencies,
     subject: Key,
-    variants: Variants
+    variants: Variants,
+    edges: &rule_graph::RuleEdges
   ) -> SelectDependencies {
+    // filters entries by whether the subject type is the right subject type
+    let dep_p_entries = edges.entries_for(
+      &vec![
+      Selector::SelectDependencies(selector.clone()),
+      Selector::select(selector.clone().dep_product)
+      ]);
+    let dep_p_entries = dep_p_entries.into_iter()
+      .filter(|e| e.matches_subject_type(subject.type_id().clone()))
+      .collect::<Vec<_>>();
+
+    let p_entries = edges.entries_for(
+      &vec![
+      Selector::SelectDependencies(selector.clone()),
+      Selector::select(selector.clone().product)
+      ]);
+
+    let p_entries = p_entries.into_iter()
```

Comment:
Wow. Was wrong. Not passing through the edges is actually more expensive time-wise. I suspect that's because doing the hash lookups per select is rather more expensive than doing it per task node.